server/src/main/java/org/opensearch/cluster/store/RemoteClusterStateService.java

```diff
@@ -0,0 +1,113 @@
+        if (blobStoreRepository == null) {
+            logger.error("Unable to set repository");
+            return;
+        }
+
+        Map<String, String> indexMetadataKeys = new HashMap<>();
+        for (IndexMetadata indexMetadata : clusterState.metadata().indices().values()) {
+            //123456789012_test-cluster/cluster-state/dsgYj10Nkso7/index/ftqsCnn9TgOX/metadata_4_1690947200
+            String indexMetadataKey = blobStoreRepository.writeIndexMetadata(clusterState.getClusterName().value(), clusterState.getMetadata().clusterUUID(),
+                indexMetadata, indexMetadataFileName(indexMetadata));
+            indexMetadataKeys.put(indexMetadata.getIndex().getName(), indexMetadataKey);
+        }
+        uploadMarker(clusterState, indexMetadataKeys);
+    }
+
+    private void setRepository() {
+        try {
+            if (blobStoreRepository != null) {
+                return;
+            }
+            if (clusterSettings.get(IndicesService.CLUSTER_REMOTE_STORE_ENABLED_SETTING)) {
+                String remoteStoreRepo = clusterSettings.get(CLUSTER_REMOTE_STORE_REPOSITORY_SETTING);
+                Repository repository = repositoriesService.get().repository(remoteStoreRepo);
+                assert repository instanceof BlobStoreRepository : "repository should be instance of BlobStoreRepository";
+                blobStoreRepository = (BlobStoreRepository) repository;
+            } else {
+                logger.info("remote store is not enabled");
+            }
+        } catch (Exception e) {
+            logger.error("set repo exception", e);
+        }
+    }
+
+    public void writeIncrementalStateAndCommit(long currentTerm, ClusterState previousClusterState, ClusterState clusterState) {
+        //todo
+    }
+
+    // why do we need this ?
+    public void writeIncrementalTermUpdateAndCommit(long currentTerm, long lastAcceptedVersion) {
+        //todo
+    }
+
+    public ClusterState getLatestClusterState(String clusterUUID) {
+        //todo
+        return null;
+    }
+
+    //todo exception handling
+    public void uploadMarker(ClusterState clusterState, Map<String, String> indexMetadataKeys) throws IOException {
+        synchronized (this) {
+            String markerFileName = getMarkerFileName(clusterState.term(), clusterState.version());
+            Map<String, ClusterMetadataMarker.UploadedIndexMetadata> uploadedIndices = indexMetadataKeys.entrySet().stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, e -> new ClusterMetadataMarker.UploadedIndexMetadata(e.getValue())));
+            ClusterMetadataMarker marker = new ClusterMetadataMarker(uploadedIndices, clusterState.term(), clusterState.getVersion(),
+                clusterState.metadata().clusterUUID(),
+                clusterState.stateUUID());
+            blobStoreRepository.writeMetadataMarker(clusterState.getClusterName().value(), clusterState.metadata().clusterUUID(), marker, markerFileName);
```

Comment:
instead of directly adding methods in blobStoreRepository, we should use `RemoteDirectory or RemoteBufferedOutputDirectory`.